src/web-worker/amp-worker.js

```diff
@@ -0,0 +1,174 @@
+  return worker.sendMessage_(method, opt_args || []);
+}
+
+/**
+ * @param {!Window} win
+ * @return {!AmpWorker}
+ * @visibleForTesting
+ */
+export function ampWorkerForTesting(win) {
+  return fromClass(win, 'amp-worker', AmpWorker);
+}
+
+/**
+ * A Promise-based API wrapper around a single Web Worker.
+ * @private
+ */
+class AmpWorker {
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const @private {!Window} */
+    this.win_ = win;
+
+    const url =
+        calculateEntryPointScriptUrl(location, 'ww', getMode().localDev);
+    /** @const @private {!Worker} */
+    this.worker_ = new win.Worker(url);
+    this.worker_.onmessage = this.receiveMessage_.bind(this);
+
+    /**
+     * Maps method names to the promise executors for in-flight invocations of
+     * those methods. E.g. messages['foo'][3] contains the {resolve, reject}
+     * functions for the third concurrent invocation of 'foo'.
+     *
+     * @const @private {
+     *   !Object<string,
+     *     !Array<({resolve: !Function, reject: !Function}|undefined)>
+     *   >
+     * }
+     */
+    this.messages_ = Object.create(null);
+  }
+
+  /**
+   * Sends a method invocation request to the worker and returns a Promise.
+   * @param {string} method
+   * @param {!Array} args
+   * @return {!Promise}
+   * @private
+   */
+  sendMessage_(method, args) {
+    const promise = new Promise((resolve, reject) => {
+      if (!this.messages_[method]) {
+        this.messages_[method] = [];
+      }
+      const index = this.messages_[method].length;
+      this.messages_[method][index] = {resolve, reject};
+
+      /** @type {ToWorkerMessageDef} */
+      const message = {method, args, id: index};
+      this.worker_./*OK*/postMessage(message);
+    });
+    return promise;
+  }
+
+  /**
+   * Receives the result of a method invocation from the worker and resolves
+   * the Promise returned from the corresponding `sendMessage_()` call.
+   * @param {!MessageEvent} event
+   * @private
+   */
+  receiveMessage_(event) {
+    const {method, returnValue, id} =
+        /** @type {FromWorkerMessageDef} */ (event.data);
+
+    // Find the stored Promise executor for this message.
+    const invocations = this.messages_[method];
+    if (!invocations) {
+      dev().error(TAG, `Received unexpected "${method}" message from worker.`);
+      return;
+    }
+
+    const resolve = invocations[id].resolve;
+    if (!resolve) {
+      dev().error(TAG, `Received unexpected "${method}" message ` +
+          `from worker with id: ${id}.`);
+      return;
+    }
+
+    resolve(returnValue);
+    invocations[id] = undefined;
+
+    // Clean up array if there are no more messages in flight for this method.
+    let empty = true;
+    for (let i = 0; i < invocations.length && empty; i++) {
+      if (invocations[i] !== undefined) {
+        empty = false;
+      }
+    }
+    if (empty) {
+      delete this.messages_[method];
```

Comment:
As explained above, this is simpler with a running index and one map independent of method name.